crates/core/src/content_store.rs

```diff
@@ -0,0 +1,40 @@
+use holochain_core_types::{entry::Entry, error::HolochainError};
+use holochain_locksmith::RwLock;
+use holochain_persistence_api::cas::{
+    content::{Address, AddressableContent, Content},
+    storage::ContentAddressableStorage,
+};
+use std::sync::Arc;
+
+pub trait ContentStore {
+    fn content_storage(&self) -> Arc<RwLock<dyn ContentAddressableStorage>>;
+}
+
+pub trait GetContent: ContentStore {
+    /// Get an entry from this storage
+    fn get(&self, address: &Address) -> Result<Option<Entry>, HolochainError> {
+        if let Some(json) = self.get_raw(address)? {
+            let entry = Entry::try_from_content(&json)?;
+            Ok(Some(entry))
+        } else {
+            Ok(None) // no errors but entry is not in chain CAS
+        }
+    }
+
+    /// Return the content at this addres, do not attempt to convert to an entry
+    fn get_raw(&self, address: &Address) -> Result<Option<Content>, HolochainError> {
+        Ok((*self.content_storage().read().unwrap()).fetch(address)?)
+    }
+
+    fn contains(&self, address: &Address) -> Result<bool, HolochainError> {
+        Ok(self.get_raw(address)?.is_some())
+    }
+}
+
+pub trait AddContent: ContentStore {
+    fn add<T: AddressableContent>(&self, content: &T) -> Result<(), HolochainError> {
```

Comment:
At the very least, this function should take a `&mut self` so we can't write to the case without a mutable reference. But with the function `content_storage()` present, it doesn't do much since anybody could just get a write lock on the CAS without using this subtrait.

Reply:
Ok I got rid of content_storage to keep all that stuff internal. It didn't add much code duplication at all really thanks to `get_raw`. And add now takes a `&mut self`